R/drive_download.R

```diff
@@ -1,60 +1,84 @@
 #' Download a file from Google Drive.
 #'
 #' @template file
+#' @param out_path Character. Path for output file. If `NULL`, we will
+#'    default to save a file to working directory by name of the Google
+#'    Drive file.
+#' @param type Character. If the file is a Google Drive type (Google Doc,
+#'    Google Sheet, Google Slide, etc.), you must specify the type of file
+#'    you would like to save (by file exension). This can be specified either
+#'    by using the `type` parameter, or by specifying the extension with the
+#'    `out_path` parameter.
+#' @param overwrite A logical scalar, do you want to overwrite a file that already
+#'    exists locally?
 #' @template verbose
 #'
 #' @return An object of class `drive_file` containing raw data from
 #'   your downloaded Google Drive file.
 #'
 #' @examples
 #' \dontrun{
-#' ## download a .csv file
-#' drive_download("chickwts.csv") %>%
-#'   write.csv("chickwts.csv")
+#' ## save "chickwts.csv" to the working directory as "chickwts.csv".
+#' drive_download(file = "chickwts.csv")
 #'
-#' ## download an .rda file
-#' drive_download("chickwts.rda") %>%
-#'   writeBin("chickwts.rda")
+#' ## save Google Sheet named "chickwts" to the working directory as
+#' ## chickwts.csv
+#' drive_download(file = "chickwts", out_file = "chickwts.csv")
 #' }
 #'
 #' @export
-drive_download <- function(file = NULL, verbose = TRUE) {
+drive_download <- function(file = NULL,
+                           out_path = NULL,
+                           type = NULL,
+                           overwrite = FALSE,
+                           verbose = TRUE) {
   file <- as_dribble(file)
   file <- confirm_single_file(file)
 
+  if (is.null(out_path)) {
+    out_path <- file$name
+  }
+
+  mime_type <- file$files_resource[[1]]$mimeType
+
+  if (!grepl("google", mime_type)) {
+
   request <- generate_request(endpoint = "drive.files.get",
                               params = list(alt = "media",
                                             fileId = file$id))
-  response <- make_request(request)
-  httr::stop_for_status(response)
-  proc_res <- httr::content(response, encoding = "UTF-8")
+  } else {
+    ## TODO make sure the type is compatable.
+    type <- type %||% tools::file_ext(out_path)
+    if (type == "") {
+      ## TODO, be able to guess type by default (for example saving sheets
+      ## as csvs..)
+      stop("We cannot guess type yet")
```

Comment:
So your plan is to use your knowledge of `mime_type` and pick a default export type here? Sounds like a good plan. Do you want to go ahead and do that while we're at this? FWIW I would consider defaulting to Excel for Sheets, because I think csv will only get you the first sheet.

This sounds like another mime type helper -- `drive_mime_export()`? -- that probably works like so:

  * `drive_mime_export()` with no input returns a tibble with ... 4 variables? Basically the "export" rows of `.drive$translate_mime_types`. It seems like we need a `default` variable where you can designate a default, for any MIME type with more than one export type (maybe that is only the Google types?). Which you would then use right here.
  * `drive_mime_export(x)` returns the default export MIME type for the input MIME type
  * `drive_mime_export(x, as = y)` checks that something of MIME type x can be exported as MIME type y. Is this useful / needed?

I can't think of any reason to NOT vectorize this on x. Is it going to need a companion for import? If so, I'd consider that too. At least make sure that the set of MIME type helper functions make sense (or will) as a group.